Improve handling of time shifts such as daylight savings

tveasey opened the pull request:
We received a data set where we had obvious deficiencies adjusting to daylight saving. There were two problems:
1. If the time series flat lines, say over a weekend, we can't necessarily segment into "before" and "after" the time shift and this was preventing us detecting the shift
2. We reverted shifts we applied sometimes where the time series is flat because in this case the test we use has low power but can have high significance.

For case 1 we simply need to check if a non-zero time shift over the full window has high significance. For case 2 we can use the same idea to revert we use when identifying whether to shift: check that we see values capturing a reasonable proportion of the time series variation before we trust the result.

I can't make the problem data set public so I haven't added unit testing, but I'll add to our regression suite. I also want to see if there are any unwanted side effects since false positives are equally important so I might iterate on this a bit.

valeriy42 commented on lib/maths/time_series/CTimeSeriesTestForChange.cc:
```diff
@@ -649,17 +704,14 @@ CTimeSeriesTestForChange::timeShift(double varianceH0,
-    // If the variance of the values is small compared to the predictor on the
-    // test interval we can't trust the shifted residual variance estimates.
-    TMeanVarAccumulator moments[2];
+    TMeanVarAccumulator valueMoments;
```
Did the comment became obsolete? Or can we update the formulation to keep the logic explained?

valeriy42 reviewed (commented):
Overall, the new changes look good.  I just have a few comments about the code documentation. Also, another unit test to cover the new logic would be great.

tveasey commented on lib/maths/time_series/CTimeSeriesTestForChange.cc:
```diff
@@ -307,7 +325,32 @@ std::string CTimeShift::print() const {
+bool CTimeShift::shouldUndo() const {
+    LOG_INFO(<< "moments = " << m_ValueMoments);
+    LOG_INFO(<< "prediction variance = " << m_PredictionVariance);
+    return common::CBasicStatistics::variance(m_ValueMoments) >=
```
I referenced the top-level comment.

tveasey commented:
I added a unit test of the change detection part alone (i.e. that we find a shift if the window doesn't contain the change). I also addressed other comments. It should be good for another pass.
